src/librustc/middle/trans/debuginfo.rs

```diff
@@ -652,41 +824,68 @@ fn declare_local(bcx: @mut Block,
     let name: &str = cx.sess.str_of(variable_ident);
     let loc = span_start(cx, span);
     let type_metadata = type_metadata(cx, variable_type, span);
-    let scope = scope_metadata(bcx.fcx, node_id, span);
 
-    let var_metadata = do name.with_c_str |name| {
-        unsafe {
-            llvm::LLVMDIBuilderCreateLocalVariable(
-                DIB(cx),
-                DW_TAG_auto_variable,
-                scope,
-                name,
-                file_metadata,
-                loc.line as c_uint,
-                type_metadata,
-                false,
-                0,
-                0)
-        }
-    };
+    let argument_index = match variable_kind {
+        ArgumentVariable(index) => index,
+        LocalVariable    |
+        CapturedVariable => 0
+    } as c_uint;
 
-    let llptr = match bcx.fcx.lllocals.find_copy(&node_id) {
-        Some(v) => v,
-        None => {
-            bcx.tcx().sess.span_bug(span, fmt!("No entry in lllocals table for %?", node_id));
+    let (var_alloca, var_metadata) = do name.with_c_str |name| {
+        match variable_access {
+            DirectVariable { alloca } => (
+                alloca,
+                unsafe {
+                    llvm::LLVMDIBuilderCreateLocalVariable(
+                        DIB(cx),
+                        DW_TAG_auto_variable,
+                        scope_metadata,
+                        name,
+                        file_metadata,
+                        loc.line as c_uint,
+                        type_metadata,
+                        cx.sess.opts.optimize != session::No,
```

Comment:
I know I'm a bit late, but was this intended?

For other functions, the `cx.sess.opts.optimize != session::No` argument corresponds to `isOptimized: bool`, but for `LLVMDIBuilderCreateLocalVariable`, it's actually `AlwaysPreserve: bool`.

And AFAICT, ever setting `AlwaysPreserve` to `false` is asking for trouble, at least when it comes to real user variables, and we don't really use LLVM debuginfo for anything else.

cc @nikomatsakis @rkruppe (I came across this while trying to get `llvm.dbg.value` to work)